arangod/Cluster/RebootTracker.cpp

```diff
@@ -0,0 +1,330 @@
+        } catch (std::exception const& ex) {
+          LOG_TOPIC("3d935", INFO, Logger::CLUSTER)
+              << "Failed to execute reboot callback: " << descr << ": " << ex.what();
+        } catch (...) {
+          LOG_TOPIC("f7427", INFO, Logger::CLUSTER)
+              << "Failed to execute reboot callback: " << descr << ": "
+              << "Unknown error.";
+        }
+      }
+    }
+  };
+}
+
+void RebootTracker::queueCallbacks(
+    std::vector<std::shared_ptr<std::unordered_map<CallbackId, DescriptedCallback>>> callbacks) {
+  if (callbacks.empty()) {
+    return;
+  }
+
+  TRI_ASSERT(std::none_of(callbacks.cbegin(), callbacks.cend(),
+                          [](auto it) { return it == nullptr; }));
+  TRI_ASSERT(std::none_of(callbacks.cbegin(), callbacks.cend(),
+                          [](auto it) { return it->empty(); }));
+
+  auto cb = createSchedulerCallback(std::move(callbacks));
+
+  if (!_scheduler->queue(RequestLane::CLUSTER_INTERNAL, std::move(cb))) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+        TRI_ERROR_QUEUE_FULL,
+        "No available threads when trying to queue cleanup "
+        "callbacks due to a server reboot");
+  }
+}
+
+void RebootTracker::unregisterCallback(PeerState const& peerState,
+                                       RebootTracker::CallbackId callbackId) {
+  MUTEX_LOCKER(guard, _mutex);
+
+  auto const cbIt = _callbacks.find(peerState.serverId());
+  if (cbIt != _callbacks.end()) {
+    auto& rebootMap = cbIt->second;
+    auto const rbIt = rebootMap.find(peerState.rebootId());
+    if (rbIt != rebootMap.end()) {
+      auto& callbackSetPtr = rbIt->second;
+      TRI_ASSERT(callbackSetPtr != nullptr);
+      callbackSetPtr->erase(callbackId);
+      if (callbackSetPtr->empty()) {
+        rebootMap.erase(rbIt);
+      }
+    }
+  }
+}
+
+RebootTracker::CallbackId RebootTracker::getNextCallbackId() noexcept {
+  _mutex.assertLockedByCurrentThread();
+
+  CallbackId nextId = _nextCallbackId;
+  ++_nextCallbackId;
+  return nextId;
+}
+
+void RebootTracker::queueCallback(DescriptedCallback callback) {
+  queueCallbacks({std::make_shared<std::unordered_map<CallbackId, DescriptedCallback>>(
+      std::unordered_map<CallbackId, DescriptedCallback>{
+          std::make_pair(getNextCallbackId(), std::move(callback))})});
+}
+
+CallbackGuard::CallbackGuard() : _callback(nullptr) {}
+
+CallbackGuard::CallbackGuard(std::function<void(void)> callback)
+    : _callback(std::move(callback)) {}
+
+// NOLINTNEXTLINE(hicpp-noexcept-move,performance-noexcept-move-constructor)
+CallbackGuard::CallbackGuard(CallbackGuard&& other)
+    : _callback(std::move(other._callback)) {
+  other._callback = nullptr;
+}
+
+// NOLINTNEXTLINE(hicpp-noexcept-move,performance-noexcept-move-constructor)
+CallbackGuard& CallbackGuard::operator=(CallbackGuard&& other) {
+  call();
+  _callback = std::move(other._callback);
+  other._callback = nullptr;
+  return *this;
+}
+
+CallbackGuard::~CallbackGuard() { call(); }
```

Comment:
What if call throws an exception? It is currently only used to call `RebootTracker::unregisterCallback`, but ...

Reply:
I'm aware of that, but couldn't decide what would be the best course of action, particularly because there currently aren't any use cases implemented. Just hiding exceptions might be bad, too.

Perhaps I should try to force the callback to be `noexcept`, WDYT?

Reply:
Forcing it to be `noexcept` would fix that problem for you... and shift it to the user of the api.

You can choose between ignoring exceptions (and logging them) or `std::terminate`.

Reply:
As we could not find a solution for the noexcept callback, I vote to leave it as it is.